File: app/src/main/java/org/zeroxlab/momodict/MainActivity.java
package org.zeroxlab.momodict;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;

import com.mikepenz.aboutlibraries.LibsBuilder;

import org.zeroxlab.momodict.ui.HistoryFragment;
import org.zeroxlab.momodict.ui.InputFragment;
import org.zeroxlab.momodict.ui.MemoFragment;
import org.zeroxlab.momodict.widget.BackKeyHandler;
import org.zeroxlab.momodict.widget.PagerFocusBroadcaster;

import java.util.ArrayList;
import java.util.List;

/**
 * Main Activity, it consists several tabs which present by ViewPager.
 */
public class MainActivity extends AppCompatActivity {

    private static final String TAG = Momodict.TAG;
    private static final int REQ_CODE_IMPORT = 0x1002;

    private ViewPager mPager;
    private MyPagerAdapter mAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_with_one_viewpager);
        initView();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        menu.add(Menu.NONE, R.id.menu_import, Menu.NONE, "Import dictionary");
        menu.add(Menu.NONE, R.id.menu_clear_history, Menu.NONE, "Clear History");
        menu.add(Menu.NONE, R.id.menu_license, Menu.NONE, "License");
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_import:
                onImportClicked();
                return true;
            case R.id.menu_clear_history:
                onClearHistory();
                return true;
            case R.id.menu_license:
                onShowLicenses();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onActivityResult(int reqCode, int resultCode, Intent data) {
        switch (reqCode) {
            case REQ_CODE_IMPORT:
                onResultImport(resultCode, data);
                break;
        }
    }

    @Override
    public void onBackPressed() {
        // 1. if user not in first tab, move to first tab
        // 2. if the first tab can handle back key, do nothing here
        // 3. otherwise, call super.onBackPressed(usually close this activity)
        if (mPager.getCurrentItem() != 0) {
            mPager.setCurrentItem(0);
        } else {
            Object first = mAdapter.getItem(0);
            if (first instanceof BackKeyHandler) {
                BackKeyHandler handler = (BackKeyHandler) first;
                if (handler.backKeyHandled()) {
                    return;
                }
            }
            super.onBackPressed();
        }
    }

    private void initView() {
        // To create fragments for Tabs, and manage them by MyPagerAdapter
        final TabLayout tabs = (TabLayout) findViewById(R.id.tabs);
        final FragmentManager mgr = getSupportFragmentManager();
        mAdapter = new MyPagerAdapter(mgr);
        mAdapter.addFragment(new InputFragment(), "Input");
        mAdapter.addFragment(new HistoryFragment(), "History");
        mAdapter.addFragment(new MemoFragment(), "Memo");

        mPager = (ViewPager) findViewById(R.id.fragment_container);
        mPager.setAdapter(mAdapter);
        // To notify fragment, when page-change happens
        mPager.addOnPageChangeListener(new PagerFocusBroadcaster(mAdapter));
        tabs.setupWithViewPager(mPager);

        //init action bar
        Toolbar toolbar = (Toolbar) findViewById(R.id.actionbar);
        toolbar.setNavigationIcon(R.mipmap.ic_logo);
        setSupportActionBar(toolbar);
    }

    /**
     * Callback when user click "Import" in Menu options
     */
    private void onImportClicked() {
        Intent i = new Intent();
        i.setClass(this, FileImportActivity.class);
        startActivityForResult(i, REQ_CODE_IMPORT);
    }

    /**
     * Callback when user click "Clear history" in Menu options
     */
    private void onClearHistory() {
        int size = mAdapter.getCount();
        for (int i = 0; i < size; i++) {
            Object o = mAdapter.getItem(i);
            if (o instanceof HistoryFragment) {
                HistoryFragment f = (HistoryFragment) o;
                f.clearHistory();
            }
        }
    }

    private void onResultImport(int resultCode, Intent data) {
        if (resultCode == Activity.RESULT_OK) {
            Uri uri = data.getData();
            Log.d(TAG, "Imported from file " + uri.getPath());
        }
    }

    /**
     * Callback when user click "License" in Menu options
     */
    private void onShowLicenses() {
        LibsBuilder builder = new LibsBuilder();
        builder.withAboutIconShown(true)
                .withAboutVersionShown(true)
                .start(this);
    }

    /**
     * Adapter for ViewPager. To use addFragment() to add items.
     */
    private class MyPagerAdapter extends FragmentStatePagerAdapter {
        final List<Fragment> iFragments = new ArrayList<>();
        final List<CharSequence> iTitles = new ArrayList<>();

        MyPagerAdapter(FragmentManager fm) {
            super(fm);
        }

        void addFragment(Fragment f, CharSequence title) {
            iFragments.add(f);
            iTitles.add(title);
        }

        @Override
        public Fragment getItem(int position) {
            return iFragments.get(position);
        }

        @Override
        public int getCount() {
            return iFragments.size();
        }

        @Override
        public String getPageTitle(int pos) {
            return iTitles.get(pos).toString();
        }
    }
}

File: app/src/main/java/org/zeroxlab/momodict/MainActivity.kt
package org.zeroxlab.momodict

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import com.google.android.material.tabs.TabLayout
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentStatePagerAdapter
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import android.util.Log
import android.util.SparseArray
import android.view.Menu
import android.view.MenuItem
import android.view.ViewGroup
import com.mikepenz.aboutlibraries.LibsBuilder
import org.zeroxlab.momodict.ui.HistoryFragment
import org.zeroxlab.momodict.ui.InputFragment
import org.zeroxlab.momodict.ui.MemoFragment
import org.zeroxlab.momodict.widget.BackKeyHandler
import org.zeroxlab.momodict.widget.PagerFocusBroadcaster
import kotlinx.android.synthetic.main.activity_with_one_viewpager.fragment_container as mPager

/**
 * Main Activity, it consists several tabs which present by ViewPager.
 */
class MainActivity : AppCompatActivity() {

    private lateinit var mAdapter: FragmentPagerAdapterImpl

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_with_one_viewpager)
        initView()
        // Should enable StrictMode for #13
        //StrictModeUtil.enableInDevMode()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_manage_dictionaries -> onManageClicked()
            R.id.menu_clear_history -> onClearHistory()
            R.id.menu_license -> onShowLicenses()
            else -> return super.onOptionsItemSelected(item)
        }
        return true
    }

    public override fun onActivityResult(reqCode: Int, resultCode: Int, data: Intent?) {
        when (reqCode) {
            REQ_CODE_IMPORT -> onResultImport(resultCode, data)
        }
    }

    override fun onBackPressed() {
        // 1. if user not in first tab, move to first tab
        // 2. if the first tab can handle back key, do nothing here
        // 3. otherwise, call super.onBackPressed(usually close this activity)
        if (mPager.currentItem != 0) {
            mPager.currentItem = 0
        } else {
            val handled = mAdapter.getFragment(0).let { firstItem ->
                when (firstItem) {
                    is BackKeyHandler -> firstItem.backKeyHandled()
                    else -> false
                }
            }

            if (!handled) {
                super.onBackPressed()
            }
        }
    }

    private fun initView() {
        // To create fragments for Tabs, and manage them by FragmentPagerAdapterImpl
        mAdapter = FragmentPagerAdapterImpl(supportFragmentManager)

        mPager.apply {
            adapter = mAdapter
            // To notify fragment, when page-change happens
            addOnPageChangeListener(PagerFocusBroadcaster(mAdapter))
        }

        with(findViewById(R.id.tabs) as TabLayout) {
            setupWithViewPager(mPager)
        }

        //init action bar
        (findViewById(R.id.actionbar) as Toolbar).let {
            it.setNavigationIcon(R.mipmap.ic_logo)
            setSupportActionBar(it)
        }
    }

    /**
     * Callback when user click "Import" in Menu options
     */
    private fun onManageClicked() {
        Intent().let {
            it.setClass(this, ManageDictActivity::class.java)
            startActivityForResult(it, REQ_CODE_IMPORT)
        }
    }

    /**
     * Callback when user click "Clear history" in Menu options
     */
    private fun onClearHistory() {
        (0..(mAdapter.count - 1))
                .map { mAdapter.getFragment(it) }
                .filterIsInstance<HistoryFragment>()
                .forEach { it.clearHistory() }
    }

    private fun onResultImport(resultCode: Int, data: Intent?) {
        when (resultCode) {
            Activity.RESULT_OK -> Log.d(TAG, "Imported from file " + data?.data?.path)
        }
    }

    /**
     * Callback when user click "License" in Menu options
     */
    private fun onShowLicenses() {
        LibsBuilder()
                .withAboutIconShown(true)
                .withAboutVersionShown(true)
                .start(this)
    }

    /**
     * Adapter for ViewPager. To use addFragment() to add items.
     */
    inner class FragmentPagerAdapterImpl internal constructor(fm: FragmentManager)
        : FragmentStatePagerAdapter(fm) {

        internal val TITLE_INPUT = "input"
        internal val TITLE_HISTORY = "history"
        internal val TITLE_MEMO = "memo"
        internal val iTitles = listOf(TITLE_INPUT, TITLE_HISTORY, TITLE_MEMO)

        internal val iFragments: SparseArray<Fragment> = SparseArray()

        override fun getItem(position: Int): Fragment {
            val title = iTitles.get(position)
            return when (title) {
                TITLE_INPUT -> InputFragment()
                TITLE_HISTORY -> HistoryFragment()
                TITLE_MEMO -> MemoFragment()
                else -> throw RuntimeException("cannot create corresponding fragment")
            }
        }

        override fun getCount(): Int = iTitles.size
        override fun getPageTitle(pos: Int): String = iTitles[pos].toString()

        override fun instantiateItem(container: ViewGroup, position: Int): Any {
            val frg = super.instantiateItem(container, position) as Fragment
            iFragments.put(position, frg)
            return frg
        }

        override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
            super.destroyItem(container, position, `object`)
            iFragments.remove(position)
        }

        fun getFragment(position: Int): Fragment? {
            return iFragments.get(position)
        }
    }

    companion object {
        private val TAG = Momodict.TAG
        private val REQ_CODE_IMPORT = 0x1002
    }
}
